gevent_.py

```diff
@@ -0,0 +1,111 @@
+import email.utils
+import os
+import mimetypes
+
+
+CRLF = '\r\n'
+ROOT = os.path.relpath('webroot')
+
+
+def server_(socket, address):
+    while True:
+        try:
+            response = ''
+            uri = None
+            uri_info = []
+            try:
+                uri = parse_request(socket)
+                uri = uri.lstrip("/")
+                uri_info = resolve_uri(uri)
+                response = response_ok(uri_info)
+            except (SyntaxError, IndexError):
+                response = response_error(400)
+            except TypeError:
+                response = response_error(406)
+            except NameError:
+                response = response_error(404)
+            except:
+                response = response_error(500)
+            socket.sendall(response)
+            socket.close()
+        except KeyboardInterrupt:
+            break
```

Comment:
Your `server_` function here is still holding on to some ghosts of the past.  

You've moved the loop that accumulates a request into the `parse_request` function, which means you can pass the client-connected socket that this function receives directly to `parse_request`.  That's cool.  

But you still have an outer loop in this function.  In the original, non-concurrent server, this outer loop exists to accept a new connection from a client and pass that connection in to the code that accumulates the request and responds to it.  In this concurrent version, the `gevent.server.StreamServer` is responsible for taking care of this task.  You don't need to any more.  

In fact, by trying to do so you introduce a very interesting bug. Your `server_` function is handed a client connection.  It enters the while loop and handles that request, sends back a response, and then closes the socket.  Then it loops back to the top where it starts to go through the whole process of handling the request on `socket` again.  But now `socket` has been closed.  This will raise an error.  The error will kill off this handler process, but certainly not in the way you had intended.  

Remove the outer loop and the try/except that catches `KeyboardInterrupt`.  These will not work properly.